### PR TITLE
Initial commit of Geekboards Macropad V2 keyboard

### PR DESCRIPTION
**Pull request by ruddy17**
Initial commit of Geekboards Macropad V2 keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/geekboards/macropad_v2/config.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+#define PERMISSIVE_HOLD
+#undef TAPPING_TOGGLE
+#define TAPPING_TOGGLE  3
+#define RETRO_TAPPING
+
+#ifndef __ASSEMBLER__
+void eeconfig_update_rgb_matrix(void);
+#define eeconfig_update_rgblight_current eeconfig_update_rgb_matrix
+#endif
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/config.h:94]:
```suggestion
```

File: keyboards/geekboards/macropad_v2/config.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#define DEBOUNCING_DELAY 3
+
+/* Bootmagic Lite key configuration */
+#define BOOTMAGIC_LITE_ROW 0
+#define BOOTMAGIC_LITE_COLUMN 0
+
+#define PERMISSIVE_HOLD
+#undef TAPPING_TOGGLE
+#define TAPPING_TOGGLE  3
+#define RETRO_TAPPING
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/config.h:89]:
```suggestion
```

File: keyboards/geekboards/macropad_v2/config.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#define RGB_MATRIX_STARTUP_SAT 255
+#define RGB_MATRIX_STARTUP_VAL 192
+#define RGB_MATRIX_STARTUP_SPD 30
+#endif //RGB_MATRIX_ENABLE
+
+#define RGB_DISABLE_WHEN_USB_SUSPENDED true
+#define WAIT_FOR_USB
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCING_DELAY 3
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/config.h:80]:
```suggestion
#define DEBOUNCE 3
```

File: keyboards/geekboards/macropad_v2/config.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x0483
+#define PRODUCT_ID      0xA372
+#define DEVICE_VER      0x0002
+#define MANUFACTURER    GeekBoards
+#define PRODUCT         GeekBoards Macropad v2
+#define DESCRIPTION     GeekBoards Macropad v2
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/config.h:27]:
```suggestion
```

File: keyboards/geekboards/macropad_v2/keymaps/default/config.h [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#pragma once
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/keymaps/default/config.h:1]: This file should be deleted

File: keyboards/geekboards/macropad_v2/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    debug_enable=true;
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    // If console is enabled, it will print the matrix position and status of each key pressed
+    #ifdef CONSOLE_ENABLE 
+    uprintf("SW%u %s\n", record->event.key.col+1 + (4*record->event.key.row), record->event.pressed?"pressed":"released");
+    #endif 
+    return true;
+}
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/keymaps/default/keymap.c:34]:
```suggestion

const uint16_t keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    LAYOUT(
        KC_1, KC_2, KC_3, KC_4,
        KC_5, KC_6, KC_7, KC_8
    )
};

```

File: keyboards/geekboards/macropad_v2/keymaps/via/rules.mk [hunk 7]
```diff
@@ -0,0 +1,3 @@
+EXTRAKEY_ENABLE = yes
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/keymaps/via/rules.mk:1]:
```suggestion
```

File: keyboards/geekboards/macropad_v2/macropad_v2.c [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+extern LED_TYPE rgb_matrix_ws2812_array[DRIVER_LED_TOTAL];
+#endif
+
+void suspend_power_down_kb(void) {
+    rgb_matrix_set_suspend_state(true);
+}
+
+void suspend_wakeup_init_kb(void) {
+    rgb_matrix_set_suspend_state(false);
+}
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/macropad_v2.c:64]:
```suggestion

void suspend_power_down_kb(void) {
    rgb_matrix_set_suspend_state(true);
}

void suspend_wakeup_init_kb(void) {
    rgb_matrix_set_suspend_state(false);
}
#endif
```

File: keyboards/geekboards/macropad_v2/macropad_v2.h [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "quantum.h"
+#define LAYOUT(k00, k01, k02, k03, k04, k05, k06, k07) {{ k00, k01, k02, k03}, {k04, k05, k06, k07}}
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/macropad_v2.h:19]:
```suggestion

#define LAYOUT( \
    k00, k01, k02, k03, \
    k10, k11, k12, k12  \
) { \
    { k00, k01, k02 k03 }, \
    { k10, k11, k12, k13 } \
}

```

File: keyboards/geekboards/macropad_v2/rules.mk [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+BOOTMAGIC_ENABLE = lite	# Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes	# Mouse keys
+POINTING_DEVICE_ENABLE = yes
+EXTRAKEY_ENABLE = yes	# Audio control and System control
+CONSOLE_ENABLE = yes		# Console for debug
+COMMAND_ENABLE = no    # Commands for debug and configuration
+SLEEP_LED_ENABLE = no  # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes	    # USB Nkey Rollover
+RGB_MATRIX_ENABLE = WS2812
+WS2812_DRIVER = pwm
```
Comment (zvecr) [inline keyboards/geekboards/macropad_v2/rules.mk:14]:
```suggestion

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
RGB_MATRIX_ENABLE = WS2812
WS2812_DRIVER = pwm

```

**Review by zvecr — CHANGES_REQUESTED** [review]
From a quick glance, you dont need the following files/folders
* `keyboards/geekboards/macropad_v2/boards/`
* `keyboards/geekboards/macropad_v2/bootloader_defs.h`

File: keyboards/geekboards/macropad_v2/config.h (hunk 1 above)
Comment (ruddy17) [reply keyboards/geekboards/macropad_v2/config.h:94]: This is necessary to use RGB_MATRIX with VIA, and this is the only way it works, that don't require core changes. I discussed this a couple times on discord, and nobody supported my idea to include these changes into core. This also requires a bit of wizardry in VIA config, nothing criminal though. Here is a demo: https://youtu.be/Sy0ZG9p_KW0

Comment (ruddy17) [reply keyboards/geekboards/macropad_v2/config.h:94]: If found appropriate, I would be happy to PR this into core, it is much more straigtforward, just a couple lines added in rgb_matrix.h.

File: keyboards/geekboards/macropad_v2/config.h [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x0483
+#define PRODUCT_ID      0xA372
+#define DEVICE_VER      0x0002
+#define MANUFACTURER    GeekBoards
+#define PRODUCT         GeekBoards Macropad v2
```
Comment (fauxpark) [inline keyboards/geekboards/macropad_v2/config.h:26]:
```suggestion
#define PRODUCT         Macropad v2
```

File: keyboards/geekboards/macropad_v2/info.json [hunk 12]
```diff
@@ -0,0 +1,54 @@
+{
+  "keyboard_name": "macropad_v2",
+  "url": "https://geekboards.ru/",
+  "maintainer": "ruddy17",
+  "width": 4,
+  "height": 2,
+  "layouts": {
+    "LAYOUT": {
+      "key_count": 8,
```
Comment (fauxpark) [inline keyboards/geekboards/macropad_v2/info.json:9]:
```suggestion
```

File: keyboards/geekboards/macropad_v2/keymaps/via/config.h [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#pragma once
+// place overrides here
```
Comment (fauxpark) [inline keyboards/geekboards/macropad_v2/keymaps/via/config.h:2]:
```suggestion

```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/geekboards/macropad_v2/config.h [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x0483
+#define PRODUCT_ID      0xA372
+#define DEVICE_VER      0x0002
+#define MANUFACTURER    GeekBoards
+#define PRODUCT         Macropad v2
+
+#define USB_MAX_POWER_CONSUMPTION 500
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/config.h:28]: This is the default and doesn't actually need to be defined.

File: keyboards/geekboards/macropad_v2/keymaps/via/config.h [hunk 15]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/keymaps/via/config.h:1]: These shouldn't be needed, at all, actually.

File: keyboards/geekboards/macropad_v2/keymaps/via/keymap.c [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+#include "quantum.h"
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/keymaps/via/keymap.c:17]:
```suggestion
```

File: keyboards/geekboards/macropad_v2/macropad_v2.c [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    2, 2, 2, 2, 2, 2,
+    4, 
+    2, 2, 2, 2,
+    4,
+    2, 2, 2
+} };
+
+void suspend_power_down_kb(void) {
+    rgb_matrix_set_suspend_state(true);
+}
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/macropad_v2.c:58]:
```suggestion
    suspend_power_down_user();
}
```

File: keyboards/geekboards/macropad_v2/macropad_v2.c [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+void suspend_wakeup_init_kb(void) {
+    rgb_matrix_set_suspend_state(false);
+}
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/macropad_v2.c:62]:
```suggestion
    suspend_wakeup_init_user();
}
```

File: keyboards/geekboards/macropad_v2/macropad_v2.h [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/macropad_v2.h:20]: Maybe? 
```suggestion
#define LAYOUT_ortho_2x4( \
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/geekboards/macropad_v2/bootloader_defs.h [hunk 20]
```diff
@@ -0,0 +1,7 @@
+/* Address for jumping to bootloader on STM32 chips. */
```
Comment (tzarc) [inline keyboards/geekboards/macropad_v2/bootloader_defs.h:1]: This file may not be needed any more due to the existence of:
platforms/chibios/GENERIC_STM32_F072XB/configs/bootloader_defs.h

**Review by tzarc — CHANGES_REQUESTED** [review]
Just a small query. Bunch of other changes still outstanding.

File: keyboards/geekboards/macropad_v2/bootloader_defs.h (hunk 20 above)
Comment (ruddy17) [reply keyboards/geekboards/macropad_v2/bootloader_defs.h]: removed it

**Review by tzarc — APPROVED** [review]

**Comment by ruddy17** [comment]
As https://github.com/qmk/qmk_firmware/pull/11084 has been accepted, I will remove duplicating fix from here. Please, hold this PR for a couple more days.

**Review by fauxpark — APPROVED** [review]

File: keyboards/geekboards/macropad_v2/config.h [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x0483
```
Comment (drashna) [inline keyboards/geekboards/macropad_v2/config.h:19]: It looks like this vendor ID belongs STMicroelectronics, and needs to be changed.

https://www.the-sz.com/products/usbid/index.php?v=0x0483&p=&n=

```suggestion
#define VENDOR_ID       0x6770
```
```
echo -n "GeekBoards" | shasum | cut -c1-4 | tr '[:lower:]' '[:upper:]'
6770
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/geekboards/macropad_v2/config.h (hunk 21 above)
Comment (ruddy17) [reply keyboards/geekboards/macropad_v2/config.h:19]: It was officially issued for our device via ST support exclusively for our device. It is guaranteed to be unique, so no problem.

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/geekboards/macropad_v2/config.h (hunk 21 above)
Comment (drashna) [reply keyboards/geekboards/macropad_v2/config.h:19]: verification was PMed to me, so 👍🏻

**Pull request merged**